### PR TITLE
Feature/unit testing v2

### PR DESCRIPTION
**Pull request by DavidPoliakoff**
I'm looking to expand our unit testing after seeing what it's like at codecov (https://codecov.io/gh/LLNL/GOTCHA)

To do this, I need to be able to test internals. My solution is a little janky, so I want to run it by you (that's what I want the review for). See https://github.com/LLNL/GOTCHA/compare/master...DavidPoliakoff:feature/unit_testing_v2?expand=1#diff-efbefc3afcaff896f345af3173bffe6bR132

If we're in a coverage test build, that macro means nothing. If we're in a build for anything other than coverage, that means static, and so we won't export it.

Thoughts? I'm game for alternatives.

Note that this is WIP, let me merge it if you approve that, I'm going to be going at this a little more

### CODE REVIEW

**Comment by DavidPoliakoff** [comment]
Works by me, updating

**Comment by DavidPoliakoff** [comment]
One of the bizarre problems is that in order to find out how to cause insert to be triggered without a function level test I had to write the test to see how grow works :p

**Comment by DavidPoliakoff** [comment]
Final edit before submit: I'm not going to make it a hammer test by default, unit tests are supposed to be extremely small. You'll see, by default small, by configuration it can take time though

File: src/hash.c [hunk 1]
```diff
@@ -53,10 +53,7 @@ int create_hashtable(hash_table_t *table, size_t initial_size, hash_func_t hashf
    return 0;
 }
 
-TEST_ONLY_VISIBILITY hash_data_t extract_data(struct hash_entry_t * entry){
-  return entry->data;
-}
-TEST_ONLY_VISIBILITY int insert(hash_table_t *table, hash_key_t key, hash_data_t data, hash_hashvalue_t value)
+int insert(hash_table_t *table, hash_key_t key, hash_data_t data, hash_hashvalue_t value)
```
Comment (mplegendre) [inline src/hash.c]: move insert back to being static

File: src/hash.c [hunk 2]
```diff
@@ -116,7 +113,7 @@ int destroy_hashtable(hash_table_t *table)
    return 0;
 }
 
-TEST_ONLY_VISIBILITY int lookup(hash_table_t *table, hash_key_t key, hash_entry_t **entry)
+int lookup(hash_table_t *table, hash_key_t key, hash_entry_t **entry)
```
Comment (mplegendre) [inline src/hash.c]: Move lookup back to being static

File: test/unit/gotcha_unit_tests.c [hunk 3]
```diff
@@ -274,56 +274,27 @@ Suite* gotcha_auxv_suite(){
...
-   int found_value = extract_data(lookup_holder);
-   ck_assert_msg(!look_return,"Internal error in lookup function");
-   del_return = removefrom_hashtable(&table,hashable_string);
-   ck_assert_msg(!del_return,"Internal error in element removal function");
-   ck_assert_msg((int)found_value==1337,"Inserted item could not be looked up in hash table");
-   ck_assert_msg(!destroy_hashtable(&table),"Could not destroy hashtable");
-}
-END_TEST
+#define NUM_INSERTS 8192
+#define TABLE_SIZE (NUM_INSERTS/2)
```
Comment (mplegendre) [inline test/unit/gotcha_unit_tests.c]: This is a test, might as well abuse the code and make TABLE_SIZE start very small so it repeatedly grows

File: test/unit/gotcha_unit_tests.c [hunk 4]
```diff
@@ -274,56 +274,27 @@ Suite* gotcha_auxv_suite(){
...
-   for(loop = 0;loop<60;loop++){
-     failed_insert_codes |= insert(&table,hashable_string,(void*)1337,dogs_val);
-   }  
-   ck_assert_msg(!failed_insert_codes,"Could not insert 60 items into 100 item table");
-
-
+   int loop;
+   for(loop=0;loop<NUM_INSERTS;loop++){
+     addto_hashtable(&table,&pointer_list[loop],loop);
+   }
```
Comment (mplegendre) [inline test/unit/gotcha_unit_tests.c]: This is buggy.  &pointer_list[loop] is a stack address with uninitialized data.  It's being passed as a key, which will in turn be used in strhash/strcmp.  I suppose it's not faulting because the string routines find a '\0' on the stack before running off the end, but it's still very undefined.

**Comment by DavidPoliakoff** [comment]
@mplegendre : I think I've addressed it the review

File: test/unit/gotcha_unit_tests.c [hunk 5]
```diff
@@ -221,11 +237,85 @@ Suite* gotcha_libc_suite(){
...
+////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////
+////////////////////////////////////////////////////GOTCHA Auxv Tests///////////////////////////////////////////////////////////////////////////////////
+////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////
+
+START_TEST(vdso_map_test){
+  struct link_map* hopefully_vdso = get_vdso_from_maps();
+  struct link_map* hopefully_also_vdso = get_vdso_from_auxv();
+  // intentionally not testing aliases explicitly as many systems don't implement this
+  ck_assert_msg(hopefully_vdso, "VDSO not found in maps");
+  ck_assert_msg(hopefully_also_vdso, "VDSO not found in auxv");
```
Comment (mplegendre) [inline test/unit/gotcha_unit_tests.c]: This will false-negative fail on some systems, as not every get_vdso_* mechanism works everywhere.  How about: 

struct link_map *maps_vdso = get_vdso_from_maps();
struct link_map *auxv_vdso = get_vdso_from_auxv();
struct link_map *alias_vdso = get_vdso_from_aliases();
ck_assert_msg(maps_vdso || auxv_vdso || alias_vdso, "VDSO not found in any solution");
ck_assert_msg(!maps_vdso || !auxv_vdso || maps_vdso == auxv_vdso, "VDSO Mismatch of maps and auxv");
ck_assert_msg(!maps_vdso || !alias_vdso || maps_vdso == alias_vdso, "VDSO Mismatch of maps and alias");
ck_assert_msg(!auxv_vdso || !alias_vdso || auxv_vdso == alias_vdso, "VDSO Mismatch of auxv and alias");

File: test/unit/gotcha_unit_tests.c [hunk 6]
```diff
@@ -221,11 +237,85 @@ Suite* gotcha_libc_suite(){
...
+  suite_add_tcase(s, libc_case);
+  return s;
+}
+
+////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////
+////////////////////////////////////////////////////GOTCHA Hash Tests///////////////////////////////////////////////////////////////////////////////////
+////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////
+
+#define NUM_INSERTS 16384
+#define TABLE_SIZE (NUM_INSERTS/2)
```
Comment (mplegendre) [inline test/unit/gotcha_unit_tests.c]: TABLE_SIZE unused.

File: test/unit/gotcha_unit_tests.c [hunk 7]
```diff
@@ -221,11 +237,85 @@ Suite* gotcha_libc_suite(){
...
+
+START_TEST(hash_grow_test){
+   hash_table_t table;
+   create_hashtable(&table,4,(hash_func_t) strhash, (hash_cmp_t) gotcha_strcmp);
```
Comment (mplegendre) [inline test/unit/gotcha_unit_tests.c]: How about 1 as the initial table size?  That might hit a painful corner case.

File: test/unit/gotcha_unit_tests.c [hunk 8]
```diff
@@ -221,11 +237,85 @@ Suite* gotcha_libc_suite(){
...
+   char* pointer_list[NUM_INSERTS];
+   hash_hashvalue_t dogs_val = strhash(hashable_string);  
+   int loop;
+   for(loop=0;loop<NUM_INSERTS;loop++){
+     pointer_list[loop] = (char*)gotcha_malloc(sizeof(char)*2);
+     pointer_list[loop][0] = (char)(loop+1);
+     pointer_list[loop][1] = (char)(0);
+   }
+   for(loop=0;loop<NUM_INSERTS;loop++){
+     addto_hashtable(&table,&pointer_list[loop],hashable_string);
```
Comment (DavidPoliakoff) [inline test/unit/gotcha_unit_tests.c]: I've updated this and it breaks, don't know what the contract is for lookup, so I'll wait for your feedback

**Comment by mplegendre** [comment]
Some of those compilation warnings look like they could be the cause of your travis test failures.  I'd recommend fixing those, but otherwise approving now.
